Feature/curry dictionary

ColmBhandal opened the pull request:
## Description

Added curry dictionary type and tests. This dictionary type nested dictionaries using tuples of keys and allows currying i.e. applying only a prefix of the keys or checking a prefix of keys for existence. The advantage is that prefixes can be queried with a complexity that is linear in the size of the key tuples only, but is constant w.r.t. the number of entries in the dictionary.

ColmBhandal committed:
BREAKING: Updated lang to 9.0 and fixed various bugs RE Nullability

ColmBhandal committed:
Added cast from PositiveInteger to NonnegativeInteger

ColmBhandal committed:
Curry dictionary - WIP

ColmBhandal committed:
Normalized functions

Added a version for params and IEnumerable for all applicable functions
Reordered functions in a standard way

ColmBhandal committed:
Added some more function to curried dictionary

Some are just stubs in advance of unit testing

ColmBhandal committed:
Added API constructor functions for curry dictionary

ColmBhandal committed:
Added tests for curry dictionary

ColmBhandal committed:
Implemented & fixed curry dictionary functions - tests now passing

ColmBhandal committed:
Updated changelog

ColmBhandal committed:
Merge branch 'develop' into feature/curry-dictionary

ColmBhandal committed:
Updated lang version in YML build file

C# 9 only supported by .NET 5 and above

ColmBhandal committed:
Merge branch 'feature/curry-dictionary' of https://github.com/ColmBhandal/CsharpExtras into feature/curry-dictionary

ColmBhandal committed:
Reverted lang to 8.0 and fixed some issuees

Returning null tuple instead of tuple containing default
Default causes a warning due to potential nullability

Fixed bug in FindFirstOccurrenceOfSet for ArrayExtension - indices were zero based and were indexing up to length

Note - FindFirstOccurrenceOfSet is different for OneBased vs. ZeroBased arrays
For one-based, it is inclusive of both upper and lower bounds
For zero-based, it excludes the upper bound

ColmBhandal committed:
Reverted .NET version in YAML build to work with C# 8

ColmBhandal committed:
Removed changelog entry for upgrade to v9

hctrdev commented on CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
+- [BREAKING] Updated language version to 9.0 (#82)
+- Added Curry Dictionary type (#82)
```
Move this to the "Added" section to be consistent with the CHANGELOG format.

ColmBhandal replied:
Ah, I didn't notice they were sub-sections!

hctrdev commented on CsharpExtras/Enumerable/OneBased/IOneBasedArray.cs:
```diff
@@ -36,8 +36,8 @@ public interface IOneBasedArray<TVal> : IEnumerable<TVal>
         IOneBasedArray<TVal> ZipMulti<TOther>(Func<TVal, TOther, TVal> zipper, IOneBasedArray<TOther> other, params IOneBasedArray<TOther>[] extras);
-        (int index, TVal element) FindFirstOccurrenceOfSet(ISet<TVal> set);
-        (int index, TVal element) FindFirstOccurrenceOfSet(ISet<TVal> set, int startIndex, int endIndex);
+        (int index, TVal element)? FindFirstOccurrenceOfSet(ISet<TVal> set);
```
Add this breaking change to the CHANGELOG file.

hctrdev commented on CsharpExtras/Enumerable/OneBased/OneBasedArray.cs:
```diff
@@ -43,18 +43,25 @@ public IEnumerator<TVal> GetEnumerator()
+        public (int index, TVal element)? FindFirstOccurrenceOfSet(ISet<TVal> set, int startIndex, int endIndex)
+        {
+            (int zeroBasedIndex, TVal element)? zeroBasedVal
+                = ZeroBasedEquivalent.FindFirstOccurrenceOfSet<TVal>(set, startIndex-1, endIndex);
```
Shouldn't the `endIndex` also be converted to a zero-based index? If it isn't it could go beyond the size of the zero based array.

ColmBhandal replied:
No- by design the end index is inclusive on OneBased and exclusive on zero-based arrays. So passing an end index of ``length`` to the zero-based is fine. And anyway, the zero-based method truncates arguments that are too long.

hctrdev commented on CsharpExtras/Enumerable/OneBased/OneBasedArray.cs:
```diff
@@ -43,18 +43,25 @@ public IEnumerator<TVal> GetEnumerator()
-            (int zeroBasedIndex, TVal element) = ZeroBasedEquivalent.FindFirstOccurrenceOfSet<TVal>(set);
-            int oneBasedIndex = zeroBasedIndex + 1;
-            return (oneBasedIndex, element);
+            return FindFirstOccurrenceOfSet(set, 1, Length + 1);
```
I don't think you need to add +1 to the Length.
```
 1  2  3  4 (one based)
 0  1  2  3 (zero based)
[a, b, c, d]
```
In the example above, whatever index type you use, the size is always 4. Meaning that this code will iterate from `1` to `5`. As you don't convert the `endIndex` to zero-based in the other function, in the zero-based function it iterate from `0` to `5`, where it should iterate from `0` to `3`.
This probably won't cause an error, because you also check the array size, but it seems wrong to me.

ColmBhandal replied:
You're right, the +1 was a mistake. Changed now.

hctrdev commented on CsharpExtrasTest/Map/Dictionary/Curry/CurryDictionaryTest.cs:
```diff
@@ -0,0 +1,111 @@
+namespace CsharpExtrasTest.Map.Dictionary.Curry
+{
+    [TestFixture]
+    public class LazyFunctionMapTest
```
Rename this class to match file name

ColmBhandal replied:
LOL what a fail

hctrdev commented on CsharpExtras/Map/Dictionary/Curry/CurryDictionaryRec.cs:
```diff
@@ -0,0 +1,124 @@
+
+namespace CsharpExtras.Map.Dictionary.Curry
+{
+    class CurryDictionaryRec<TKey, TVal> : CurryDictionaryBase<TKey, TVal>
```
Maybe rename this to `CurryDictionaryRecursive`? It might make it a bit clearer.

ColmBhandal replied:
Done

hctrdev commented on CsharpExtras/Map/Dictionary/Curry/ICurryDictionary.cs:
```diff
@@ -0,0 +1,66 @@
+    public interface ICurryDictionary<TKey, TVal>
+    {
+        NonnegativeInteger Arity { get; }
+        TVal this[params TKey[] keyTuple] { get; set; }
```
Is there a reason to have a `set` property here? Neither of the implementations seem to implement it.

ColmBhandal replied:
Yeah that ``set`` shouldn't be there. Good catch.

ColmBhandal committed:
Minor PR fixes

Moved changelog entry to correct location
Did some renaming
Fixed minor inconsistency in FindFirstOccurrenceOfSet by removing +1
Removed unimiplemented setter property on curry dictionary

ColmBhandal committed:
Refactored one-based find first occurrence function to make end index exclusive

ColmBhandal committed:
Fixed bug

Need to add 1 when forwarding to other overload because upper bound is now exclusive
